src/NHibernate/Type/EntityType.cs

```diff
@@ -87,6 +88,22 @@ public override bool IsEqual(object x, object y, ISessionFactoryImplementor fact
 
 		public virtual bool IsNull(object owner, ISessionImplementor session)
 		{
+			if (IsNullable && !string.IsNullOrEmpty(associatedEntityName))
+			{
+				EntityEntry entry = session.PersistenceContext.GetEntry(owner);
+				int position = -1;
+				for (int i = 0; i < entry.Persister.PropertyTypes.Length; i++)
+				{
+					if (entry.Persister.PropertyTypes[i].Name == associatedEntityName)
```

Comment:
This would not work if "owner" has more than one property referencing this entity.